x/oracle/keeper/keeper.go

```diff
@@ -147,23 +147,16 @@ func (k Keeper) ProcessClaim(ctx sdk.Context, claim types.Claim) (types.Prophecy
 	return prophecy, nil
 }
 
-func (k Keeper) checkActiveValidator(ctx sdk.Context, validatorAddress sdk.ValAddress) bool {
-	validator, found := k.stakeKeeper.GetValidator(ctx, validatorAddress)
-	if !found {
-		return false
-	}
-
-	return validator.GetStatus().Equal(sdk.Bonded)
-}
-
 // processCompletion looks at a given prophecy
 // and assesses whether the claim with the highest power on that prophecy has enough
 // power to be considered successful, or alternatively,
 // will never be able to become successful due to not enough validation power being
 // left to push it over the threshold required for consensus.
 func (k Keeper) processCompletion(ctx sdk.Context, prophecy types.Prophecy) types.Prophecy {
-	highestClaim, highestClaimPower, totalClaimsPower := prophecy.FindHighestClaim(ctx, k.stakeKeeper)
-	totalPower := k.stakeKeeper.GetLastTotalPower(ctx)
+	highestClaim, highestClaimPower, totalClaimsPower, totalPower := prophecy.FindHighestClaim(ctx, k.stakeKeeper)
+	if !sdk.IsUpgrade(sdk.BEP159Phase2) {
+		totalPower = k.stakeKeeper.GetLastTotalPower(ctx)
```

Comment:
Just curious why have to fetch `totalPower` again?

Reply:
In the previous implementation, we get validators' power by `keeper.GetBondedValidatorsByPower` and total power by `keeper.GetLastTotalPower`.

In the new implementation, since we introduce whitelist relayers, we snapshot mapping of whitelist relayer addresses to their power, thus we can sum the powers in the snapshot up to get the total power.

We fetch `totalPower` here again before BEP159Phase2 to remain backward compatibility.

It is a little bit weird to fetch it here indeed. I will move it into `FindHighestClaim` function.